homeassistant/components/eq3btsmart/__init__.py

```diff
@@ -28,47 +24,38 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-type Eq3ConfigEntry = ConfigEntry[Eq3ConfigEntryData]
-
-
 async def async_setup_entry(hass: HomeAssistant, entry: Eq3ConfigEntry) -> bool:
     """Handle config entry setup."""
-
-    mac_address: str | None = entry.unique_id
-
-    if TYPE_CHECKING:
-        assert mac_address is not None
-
-    eq3_config = Eq3Config(
-        mac_address=mac_address,
-    )
+    mac_address: str = entry.data.get(CONF_MAC_ADDRESS, entry.unique_id)
 
     device = bluetooth.async_ble_device_from_address(
         hass, mac_address.upper(), connectable=True
     )
 
     if device is None:
-        raise ConfigEntryNotReady(
-            f"[{eq3_config.mac_address}] Device could not be found"
+        raise ConfigEntryNotReady(f"[{mac_address}] Device could not be found")
+
+    if not entry.data:
+        hass.config_entries.async_update_entry(
+            entry,
+            data={CONF_MAC_ADDRESS: mac_address},
         )
 
     thermostat = Thermostat(device)
+    coordinator = Eq3Coordinator(hass, entry, mac_address, thermostat)
+    entry.runtime_data = Eq3ConfigEntryData(thermostat, coordinator)
 
-    entry.runtime_data = Eq3ConfigEntryData(
-        eq3_config=eq3_config, thermostat=thermostat
-    )
     entry.async_on_unload(entry.add_update_listener(update_listener))
+
+    await coordinator.async_config_entry_first_refresh()
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
-    entry.async_create_background_task(
-        hass, _async_run_thermostat(hass, entry), entry.entry_id
-    )
+    coordinator.async_update_listeners()
```

Comment:
so the thing with coordinator, you should make sure your entities can infer the state directly, without needing the call to update the listeners